miniupnpd/pf/obsdrdr.c

```diff
@@ -1471,11 +1471,13 @@ get_redirect_rule_by_index(int index,
 	}
 #else /* USE_LIBPFCTL */
 	pr.nr = index;
-	if(ioctl(dev, DIOCGETRULE, &pr) < 0)
-	{
-		syslog(LOG_ERR, "ioctl(dev, DIOCGETRULE): %m");
-		goto error;
-	}
+	do {
+		if(ioctl(dev, DIOCGETRULE, &pr) < 0)
+		{
+			syslog(LOG_ERR, "ioctl(dev, DIOCGETRULE): %m");
+			goto error;
+		}
+	} while ( pr.nr != (unsigned int)index );
```

Comment:
I'm confused : is there a risk of infinite loop ?
and why pr.nr would be different in the first place ?

Reply:
https://man.openbsd.org/pf has not changed : 

> [DIOCGETRULE](https://man.openbsd.org/pf#DIOCGETRULE) _struct pfioc_rule *pr_
>     Get a _rule_ by its number _nr_ using the ticket obtained through a preceding `DIOCGETRULES` call. If action is set to `PF_GET_CLR_CNTR`, the per-rule statistics on the requested _rule_ are cleared.

Reply:
No, the issue is that it will always return rules in order. So, if the loop is not there it will always return the first rule when it is called (setting pr.nr before has no effect). The loop is there so that it will actually return the requested index.

Reply:
As an example, say there are 4 rules in the anchor. The very first rule is one that has days left before it expires (since a week is the max). The other rules should expire within the next hour. Calling this will _always_ return the first rule (pr.nr == 0) for all calls to `get_redirect_rule_by_index()` without this change (on OpenBSD 7.5, at least). This breaks the other rules being cleaned when expired, as only the very first rule in the list will ever be checked.

Reply:
There should be no functional change for any system where setting pr.nr before calling DIOCGETRULE works. In that case the test should fail and it will never loop.

Reply:
Oh I missed your question about if it would be an infinite loop. It shouldn't, if `DIOCGETRULE` returns an error (like getting to the end of the list), the `goto` statement will break it out of the loop.

Reply:
What's strange is that the current OpenBSD man states that setting pr.nr before calling DIOCGETRULE works.
Well at least it was the way it worked when they wrote the man and they haven't changed the man.

That looks like a bug in either OpenBSD man or the DIOCGETRULE implementation ;)

Reply:
I'll agree with you there. A bug in one or the other. This is also why I'm not sure when it changed. Possibly 7.4, when other pf stuff was changed (like the addition of DIOCXEND).

Reply:
7.4 may indeed be when it changed. I was looked at #674 where support for `DIOCXEND` was added, and I had this from the 7.4 changelog:

```
Speed up the ioctl(2) request DIOCGETRULE such that pfctl(8)
can retrieve all pf(4) rules from the kernel in linear rather than
in quadratic time. To protect the kernel from memory exhaustion,
userland processes now have to release tickets obtained with
DIOCGETRULES by issuing the new ioctl(2) request DIOCXEND.
In particular, snmpd(8) and systat(1) now do that.
```

That bit about speeding up `DIOCGETRULE` wasn't really relevant to that pull request, it was just part of the same block that had the bit about `DIOCXEND`, but I'll bet this was when it was changed.

Reply:
I found the change : https://github.com/openbsd/src/commit/072583c99019a91ac511a1e33e10ad02df79ec97
indeed after this commit `pr.nr` is not an input anymore, but an output !